src/main/kotlin/io/github/inductiveautomation/kindling/gatewaynetwork/GatewayNetworkViewer.kt

```diff
@@ -0,0 +1,161 @@
+                    if (desktop != null && desktop.isSupported(Desktop.Action.BROWSE)) {
+                        try {
+                            desktop.browse(URI(tempAddress))
+                        } catch (error: Exception) {
+                            JOptionPane.showMessageDialog(
+                                null,
+                                "Error: " + error.message,
+                                "Error opening browser",
+                                JOptionPane.ERROR_MESSAGE,
+                            )
+                        }
+                    }
+                }
+            },
+        )
+
+        // Temp file cleanup when Kindling exits.
+        val shutdownThread = Thread {
+            val tmpDir: File = Path.of(System.getProperty("java.io.tmpdir"), "gateway-network-diagram").toFile()
+            tmpDir.deleteRecursively()
+        }
+
+        Runtime.getRuntime().addShutdownHook(shutdownThread)
+    }
+
+    companion object {
+        private val FAVICON = "favicon.ico"
+        private val INDEX_HTML = "index.html"
+        private val MAIN_JS = "main.js"
+        private val STYLE_CSS = "style.css"
+        private val EXTERNAL_DIAGRAM_JS = "external-diagram.js"
+        private val PREAMBLE = "window.externalDiagram = "
+    }
+
+    private fun parseLines(lines: Sequence<String>, textArea: JTextArea) {
+        val builder = StringBuilder()
+        for (line in lines) {
+            if (line.isBlank()) {
+                continue
+            } else {
+                builder.append(line)
+                builder.append("\n")
+            }
+        }
+        textArea.setText(builder.toString())
+    }
+
+    private fun writeStaticFiles(jsonText: String): String {
+        // Make a temp dir for the static html/js files
+        val tmpDir: Path = Path.of(
+            System.getProperty("java.io.tmpdir"),
+            "gateway-network-diagram",
+            UUID.randomUUID().toString(),
+        )
+        Files.deleteIfExists(tmpDir)
+        Files.createDirectories(tmpDir)
+
+        val favicon: URL? = GatewayNetworkViewer::class.java.getResource(FAVICON)
+        Files.copy(Path.of(favicon?.toURI()), tmpDir.resolve(FAVICON))
+
+        val indexHtml: URL? = GatewayNetworkViewer::class.java.getResource(INDEX_HTML)
+        Files.copy(Path.of(indexHtml?.toURI()), tmpDir.resolve(INDEX_HTML))
+
+        val mainJs: URL? = GatewayNetworkViewer::class.java.getResource(MAIN_JS)
+        Files.copy(Path.of(mainJs?.toURI()), tmpDir.resolve(MAIN_JS))
+
+        val styleCss: URL? = GatewayNetworkViewer::class.java.getResource(STYLE_CSS)
+        Files.copy(Path.of(styleCss?.toURI()), tmpDir.resolve(STYLE_CSS))
+
+        val externalDiagramFile = tmpDir.resolve(EXTERNAL_DIAGRAM_JS)
+        Files.writeString(externalDiagramFile, PREAMBLE + jsonText)
+        return tmpDir.toString().replace("\\", "/")
```

Comment:
I wonder if there's a better type to return here than String - can we convert the path to a URI more directly, or assume the protocol somehow?